### PR TITLE
not crash on empty list

### PR DESCRIPTION
**Pull request by strake**

### CODE REVIEW

**Comment by strake** [comment]
ping

File: src/Z3/Base.hs [hunk 1]
```diff
@@ -1034,15 +1033,15 @@ mkBool ctx True  = mkTrue  ctx
 
 -- | Create an AST node representing args[0] + ... + args[num_args-1].
 mkAdd :: Context -> [AST] -> IO AST
-mkAdd = liftAstN "Z3.Base.mkAdd: empty list of expressions" z3_mk_add
+mkAdd ctx = maybe (mkInteger ctx 0) (liftAstN1 z3_mk_add ctx) . nonEmpty
```
[the remark this answers is not in the document]
Comment (strake) [reply src/Z3/Base.hs]: They can, but it feels cumbersome:
```haskell
mkAdd cxt = liftAstN (pure $ mkInteger cxt 0 >>= flip h2c pure) z3_mk_add cxt
```
Let me know whether you prefer this. (Maybe i'm missing a more obvious cleaner definition.)

File: src/Z3/Base.hs [hunk 2]
```diff
@@ -987,9 +989,8 @@ mkEq = liftFun2 z3_mk_eq
 -- distinct.
 --
 -- That is, @and [ args!!i /= args!!j | i <- [0..length(args)-1], j <- [i+1..length(args)-1] ]@
-mkDistinct :: Context -> [AST] -> IO AST
-mkDistinct =
-  liftAstN "Z3.Base.mkDistinct: empty list of expressions" z3_mk_distinct
+mkDistinct :: Context -> NonEmpty AST -> IO AST
```
[the remark this answers is not in the document]
Comment (strake) [reply src/Z3/Base.hs]: Done, with a pragma to warn users of failures, let me know whether you want to keep it

**Comment by strake** [comment]
@IagoAbal You said in another thread you mean to cut a release tomorrow. Think you might have time to review this latest commit before then?

**Comment by IagoAbal** [comment]
I will most likely take this, or a variation of this, but I'm trying to decide whether which pair of functions should be the default ones. I am not sure whether the use of `NonEmpty` is popular enough to make it the default. On the other hand, one could also add an extra `default` argument to provide a value in case the list is empty, or return `Maybe`, or perhaps even go a bit further and make `MonadZ3` to derive from `MonadError` or similar.

I think a `NonEmpty` variant is nice if you are already using `NonEmpty` in your program, but if you're using lists, then you'll have to use `NonEmpty.fromList` to convert them, and I think this will also throw an exception if the list is non-empty, so it won't help much. So I guess I'm already leaning towards adding these functions as a complement, but not to replace the existing ones.

Thoughts?

**Comment by strake** [comment]
I strongly prefer this formulation as written to those alternatives:
• Returning `Maybe` requires checking the value, even if i do have a `NonEmpty` to pass
• Taking a default argument is equivalent to returning `Maybe` (essentially including the `Maybe` eliminator in the function)
• Declaring `class MonadError ... => MonadZ3 ...` will not help users of the other APIs, and ultimately has the same problem as returning `Maybe`, but for **every** function in the API

Taking `NonEmpty` would indeed require inserting a check if the caller is using `[]`, but returning `Maybe` would require a check in either case; taking `NonEmpty` would leave the choice of whether to encode the non-emptiness in the type, and what error message to raise in the empty case, to the user, at the (in my opinion) small cost in verbosity of writing `fromList` somewhere.

That said, i do think your idea in #16 (i.e. to gather breaking modifications into a single breaking version bump) is a good one.

**Comment by IagoAbal** [comment]
I merged a slightly modified version of your patch in commit a0ed32b2cd002b479885464384eca7a8baf3ed20, so that backwards compatibility is preserved.

Could you give v408.0 a try and let me know if it works for you? I would then upload it to Hackage.

**Pull request closed without merging**